Improve std.exception.handle

JakobOvrum opened the pull request:
* Refactored to not use AliasThis as discussed in #2949
* Added `RangePrimitive.access` and `RangePrimitive.pop` to make doing the right thing in the common case easier
* Added example using `handle` to replace invalid code points with spaces instead of throwing

Ping @burner; I examined the possibility of adding `handleAccess` and `handlePop` convenience functions, but I think this way, using additional members of `RangePrimitive` is cleaner. IMO, this way, doing the right thing is easy while flexibility is still available. What do you think?

About opSlice infinite ranges, maybe leaving the `take`ing to the handler function and static-asserting if the return type isn't correct would be a nice solution. That way the user can't make the easy mistake of not implementing slicing properly, but still has the flexibility of returning a slice of whatever length the user wants.

burner commented:
I'm not against adding handlePop.
I writing a patch to pass the size_t idx of opSlice to the handler. You can't really handle what went wrong if you don't have the index. The same is true for opSlice. You need to now what the indices where in order to fix the cause of the exception. I know that would create three kinds of handler, IMO that's is worth the trouble.

burner commented:
whats wrong with the alias this?

JakobOvrum commented:
> whats wrong with the alias this?

See #2949.

> I writing a patch to pass the size_t idx of opSlice to the handler. You can't really handle what went wrong if you don't have the index. The same is true for opSlice. You need to now what the indices where in order to fix the cause of the exception. I know that would create three kinds of handler, IMO that's is worth the trouble.

I think receiving the index (and lower/upper bounds for opSlice) should be optional; much like the ConvException example you wrote, the handler wanting to return a simple sentinel value independent of the index is probably a very common case.

burner commented:
I get the alias this problem. 
I'm also fine with the optional size_t parameter for opIndex.
opSlice should also have the optional parameter version.

JakobOvrum commented:
> I'm also fine with the optional size_t parameter for opIndex.
> opSlice should also have the optional parameter version.

Were you working on a patch for this, or do you want me to add it to this PR? Either way works for me.

burner commented:
I thought you were already working on it.

JakobOvrum commented:
> I thought you were already working on it.

`opIndex` is easy, but it turns out that it's rather murky for `opSlice` because of the idea of the dollar token type for ranges without length. I added support for the optional index argument to `opIndex` but I'll leave support for `opSlice` outside the scope of this PR.

I also added some function attributes to the unit tests to verify purity, nothrow and memory safety where appropriate.

burner commented:
LGTM
yeah opSlice is a beast.
Let me know if you got anything else to add, if not I will merge.

Thanks

monarchdodra commented:
> Let me know if you got anything else to add, if not I will merge.

You are part of the Phobos team now? Nice. I completely missed that. Congrats.

burner commented:
thanks

JakobOvrum commented:
> Let me know if you got anything else to add, if not I will merge.

I'm done for now, unless you have any more good ideas :)

quickfur commented:
Let's merge. Future enhancements can be done as followup PR's.

quickfur commented:
Auto-merge toggled on
